email/reporter.go

```diff
@@ -0,0 +1,231 @@
+		logger.Errorf("Failed fetching unreported emails, error %v", err)
+		return
+	}
+
+	// log unreported message count
+	numUnreported := len(toReport)
+	if numUnreported == 0 {
+		logger.Debugf("Found %v unreported messages", numUnreported)
+		return
+	}
+
+	logger.Infof("Found %v unreported messages", numUnreported)
+
+	// loop all emails and report them
+	for _, email := range toReport {
+		err := r.reportEmail(email)
+		if err != nil {
+			logger.Errorf("Failed to report email %v, error %v", email.UID, err)
+		}
+	}
+}
+
+// reportEmail will report the given email to NCMEC, it does so by interacting
+// with the NCMEC API.
+func (r *Reporter) reportEmail(email database.AbuseEmail) error {
+	// convenience variables
+	logger := r.staticLogger
+
+	// convert the abuse email into a NCMEC report
+	report, err := emailToReport(email)
+	if err != nil {
+		return err
+	}
+
+	// open the report
+	reportedAt := time.Now().UTC()
+	resp, err := r.staticClient.openReport(report)
+	if err == nil && resp.ResponseCode != 0 {
+		err = fmt.Errorf("unexpected response code %v when opening report for email '%v'", resp.ResponseCode, email.ID.Hex())
+	}
+	if err != nil {
+		logger.Errorf("failed to open report, err '%v'", err)
+		return err
+	}
+	reportId := resp.ReportId
+
+	// defer an update to the email with some information about the NCMEC report
+	var reportErr error
+	defer func() {
+		var reportErrStr string
+		if reportErr != nil {
+			reportErrStr = reportErr.Error()
+		}
+		if err := r.staticDatabase.UpdateNoLock(email, bson.D{
+			{"$set", bson.D{
+				{"reported", reportErr == nil},
+				{"reported_at", reportedAt},
+
+				{"ncmec_report_id", reportId},
+				{"ncmec_report_err", reportErrStr},
+			}},
+		}); err != nil {
+			logger.Errorf("failed to update email %v with report id %v, err '%v'", email.MessageID, reportId, err)
+		}
+	}()
+
+	// finish the report
+	var reportRes reportDoneResponse
+	reportRes, reportErr = r.staticClient.finishReport(reportId)
+	if reportErr == nil && reportRes.ResponseCode != 0 {
+		reportErr = fmt.Errorf("unexpected response code %v when finishing report for email '%v'", resp.ResponseCode, email.ID.Hex())
+	}
+	if reportErr != nil {
+		logger.Errorf("failed to finish report %v, err '%v'", reportId, reportErr)
+		return reportErr
+	}
+
+	return nil
+}
+
+// emailToReports returns a report containing every unique skylink in the
+// original email report, note that only csam links will be turned into a report
+// to NCMEC.
+func emailToReport(email database.AbuseEmail) (report, error) {
+	// sanity check it's parsed
+	if !email.Parsed {
+		return report{}, errors.New("email has to be parsed")
+	}
+
+	// sanity check it's csam
+	pr := email.ParseResult
+	if !pr.HasTag("csam") {
+		return report{}, errors.New("email has to contain csam")
+	}
+
+	// construct the urls
+	urls := make([]string, len(pr.Skylinks))
+	for i, skylink := range pr.Skylinks {
+		urls[i] = fmt.Sprintf("https://siasky.net/%s", skylink)
```

Comment:
The portal URL needs to be configurable.

Reply:
Good point. We should already have an env var for that anyway.

Reply:
Made all of these configurable, was going to F/U with that but did it here.